### PR TITLE
Implement IXmlRepository storing documents in Google Cloud Storage

### PR DESCRIPTION
**Pull request by jskeet**
This has no *unit* tests as the integration tests are much more useful. Suggestions for unit tests welcome, of course.

It's not clear how ASP.NET Core supports the case where multiple servers add an element for the same purpose at the same time, but that's not something this code can affect.

### CODE REVIEW

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                    attempt++;
+                    int millisecondsToSleep;
+                    lock (_lock)
+                    {
+                        int nextBackoffMillis = (int)nextBackoff.TotalMilliseconds;
+                        // Apply jitter to the backoff, but only within the range of 50%-100% of the "theoretical" backoff.
+                        millisecondsToSleep = nextBackoffMillis / 2 + _random.Next(nextBackoffMillis);
+                    }
+                    Thread.Sleep(millisecondsToSleep);
+                    nextBackoff = TimeSpan.FromSeconds(nextBackoff.Seconds * _backoffMultiplier);
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: `nextBackoff.Seconds` should be `.TotalSeconds`

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+using System;
+using System.Threading;
+
+namespace Google.Cloud.AspNetCore.DataProtection.Storage
+{
+    /// <summary>
+    /// Simple means of executing retry with exponential backoff and proportional jitter,
+    /// retrying on any GoogleApiException.
+    /// </summary>
+    internal sealed class RetryHandler
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: Don't we already have code that does this??

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+                        // Apply jitter to the backoff, but only within the range of 50%-100% of the "theoretical" backoff.
+                        millisecondsToSleep = nextBackoffMillis / 2 + _random.Next(nextBackoffMillis);
+                    }
+                    Thread.Sleep(millisecondsToSleep);
+                    nextBackoff = TimeSpan.FromSeconds(nextBackoff.Seconds * _backoffMultiplier);
+                    if (nextBackoff > _maxBackoff)
+                    {
+                        nextBackoff = _maxBackoff;
+                    }
+                    continue;
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: `continue` not really needed.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                }
+                catch (GoogleApiException) when (attempt < _maxAttempts)
+                {
+                    attempt++;
+                    int millisecondsToSleep;
+                    lock (_lock)
+                    {
+                        int nextBackoffMillis = (int)nextBackoff.TotalMilliseconds;
+                        // Apply jitter to the backoff, but only within the range of 50%-100% of the "theoretical" backoff.
+                        millisecondsToSleep = nextBackoffMillis / 2 + _random.Next(nextBackoffMillis);
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: Won't this produce a range of 50%-150%, which disagrees with the comment?

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+                if (latest.Generation != 0 && latest.Document == null)
+                {
+                    // This will throw immediately, with no retry.
+                    throw new InvalidOperationException("Object does not contain XML.");
+                }
+
+                XDocument document = latest.Generation == 0 ? new XDocument(new XElement("root")) : new XDocument(latest.Document);
+                document.Root.Add(element);
+
+                //File.AppendAllText("store.txt", $"{DateTime.UtcNow:HH:mm:ss.fff}: Trying to append {element.Attribute("id")} to {_objectName}\r\n");
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Remove commented-out code.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+            lock (_lock)
+            {
+                _latestDocument = current;
+            }
+            return current;
+        }
+
+        /// <summary>
+        /// Loads the latest document, based on a possible "currently known latest" document.
+        /// This method does not retry. Note that this could be a partial method, but it's simpler to
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: nit: Do you mean "nested method"?

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+                stream.Position = 0;
+                XDocument document = null;
+                try
+                {
+                    document = XDocument.Load(stream);
+                }
+                catch (XmlException)
+                {
+                    // The element variable will still be null, which is fine.
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Do you mean the `document` variable?

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+                }
+                return new LoadDocumentResult(nextGeneration, document);
+            }
+        }
+
+        /// <summary>
+        /// Information from the last time we loaded the document. There are three states:
+        /// - Document doesn't exist (Generation is 0, Element is null)
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: `Element` -> `Document`?

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            Assert.Equal("id1", element.Attribute("id").Value);
+        }
+
+        [Fact]
+        public void NonEmptyInitialXmlDocument()
+        {
+            string objectName = nameof(NonEmptyInitialXmlDocument);
+            UploadObject(objectName, "<differentroot><other id=\"old\" /></differentroot>", "text/xml");
+            var repo = CreateRepo(objectName);
+            XElement element = Assert.Single(repo.GetAllElements());
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs]: `element` is currently unused.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+            var allIds = document.Root.Elements().Select(x => x.Attribute("id").Value).OrderBy(id => id).ToList();
+            expectedIds = expectedIds.OrderBy(id => id).ToList();
+            Assert.Equal(expectedIds, allIds);
+        }
+
+        private XElement CreateElement(string id) => new XElement("element", new XAttribute("id", id));
+
+        /// <summary>
+        /// Starts a task which will wait for the given semaphore to be released, then
+        /// adds the given item to the given repository. This uses a thread rather than a task to ensure
```
Comment (chrisdunelm) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs]: I don't understand how this is ensuring that a single thread isn't being used?

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs (hunk 1 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: Ooh! That may actually explain the integration tests failing with relatively few changes. Thank you!

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs (hunk 2 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: Sort of - but in the Gax.Grpc package, which we're not using here :( And I don't want to end up taking a transitive dependency on it for that reason.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs (hunk 4 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/RetryHandler.cs]: Fixed, thanks (to include another /2). It was all fine, and then I refactored badly...

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs (hunk 5 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Done

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs (hunk 6 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Yes. Updated.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs (hunk 7 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Yup. It was element at one point...

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs (hunk 8 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Done.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs (hunk 10 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.IntegrationTests/DataProtection.Storage/CloudStorageXmlRepositoryTest.cs]: Comment was written when it was going to return Thread. I fail.
Have added a ConfigureAwait(false) so things really will be running on the thread pool. We *could* use Task.Factory.StartNew with a "long-running", but that's probably overkill.

**Comment by jskeet** [comment]
Thanks for the careful review! Will try bumping up the concurrency stress tests after fixing the broken retry...

**Comment by jskeet** [comment]
Added a commit to address the comments, but now one of the tests is failing. Looking into that more.

**Comment by jskeet** [comment]
Wahoo :) Will rebase to a single commit, force push, then merge on green. Thanks folks!

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+using System.Net;
+using System.Xml;
+using System.Xml.Linq;
+
+using Object = Google.Apis.Storage.v1.Data.Object;
+
+namespace Google.Cloud.AspNetCore.DataProtection.Storage
+{
+    /// <summary>
+    /// Implementation of <see cref="IXmlRepository"/> that stores the protected elements in a single Google Cloud Storage file.
```
Comment (SurferJeffAtGoogle) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Protected?  My understanding was that the IXmlRepository class is that it's a general class for storing XML.  The [docs](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.dataprotection.repositories.ixmlrepository?view=aspnetcore-2.2) say simply _The basic interface for storing and retrieving XML elements._ We happen to use the implementation to store encrypted stuff, but I see nothing about the interface that states or implies protection.

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+            initialBackoff: TimeSpan.FromSeconds(1),
+            backoffMultiplier: 2.0,
+            maxBackoff: TimeSpan.FromSeconds(3));
+
+        private readonly StorageClient _client;
+        private readonly string _bucketName;
+        private readonly string _objectName;
+
+        private readonly object _lock = new object();
+        private LoadDocumentResult _latestDocument;
```
Comment (SurferJeffAtGoogle) [inline Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Keeping _latestDocument adds a lot of complexity to the code, and I'm not sure of the benefit.
In what sequence of calls will caching the the latest document yield a benefit?

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs (hunk 11 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Well the namespace is Microsoft.AspNetCore.DataProtection.Repositories so I assume it's to do with data protection. Happy to change it if we ever want. (This is an internal comment.)

File: Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs (hunk 12 above)
Comment (jskeet) [reply Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.DataProtection.Storage/CloudStorageXmlRepository.cs]: Anything that calls GetElements many times, which I think may happen. It still makes a metadata request per call, but doesn't fetch the data.

**Pull request merged**